lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -245,6 +246,87 @@ def build_authorization_request(money, payment_method, options)
         end
       end
 
+      def add_additional_data(xml, amount, payment_method, options)
+        xml.branchSpecificExtension do
+          xml.tag! 'purchase' do
+            add_level_two_data(xml, amount, payment_method, options)
+            add_level_three_data(xml, amount, payment_method, options)
+          end
+        end
+      end
+
+      def add_level_two_data(xml, amount, payment_method, options)
+        xml.invoiceReferenceNumber options[:invoice_reference_number] if options[:invoice_reference_number]
+        xml.customerReference options[:customer_reference] if options[:customer_reference]
+        xml.cardAcceptorTaxId options[:card_acceptor_tax_id] if options[:card_acceptor_tax_id]
+        xml.tag! 'salesTax' do
+          add_amount(xml, options[:sales_tax][:amount].to_i, options[:sales_tax]) if options[:sales_tax]
+        end
+      end
+
+      def add_level_three_data(xml, amount, payment_method, options)
+        if options[:discount_amount]
+          xml.discountAmount do
+            add_amount(xml, options[:discount_amount][:amount].to_i, options[:discount_amount])
+          end
+        end
+        if options[:shipping_amount]
+          xml.shippingAmount do
+            add_amount(xml, options[:shipping_amount][:amount].to_i, options[:shipping_amount])
+          end
+        end
+        if options[:duty_amount]
+          xml.dutyAmount do
+            add_amount(xml, options[:duty_amount][:amount].to_i, options[:duty_amount])
+          end
+        end
+        if options.include?(:item)
+          item = options[:item]
+          xml.tag! 'item' do
+            xml.description item[:description] if item[:description]
+            xml.productCode item[:product_code] if item[:product_code]
+            xml.commodityCode item[:commodity_code] if item[:commodity_code]
+            xml.quantity item[:quantity] if item[:quantity]
+
+            if item[:unit_cost]
```

Comment:
💬  Comment:

Same here, I thing is a better idea to find a generic way to add those fields with out all those if statements, something like:

```Ruby
{
  unit_cost: "unitCost",
  item_total: "itemTotal",
  item_total_with_tax: "itemTotalWithTax",
  item_discount_amount: "itemDiscountAmount",
  tax_amount: "taxAmount"
}.each do |key, tag|
  xml.tag! tag  do
    add_amount(xml, item[key][:amount].to_i, item[key])
  end if item[key]
end
```